pkg/orchestration/wiring/datadog/plugin.go

```diff
@@ -0,0 +1,201 @@
+	ResourceType                  voyager.ResourceType = "datadog-alarm"
+)
+
+type WiringPlugin struct {
+	svccatentangler.SvcCatEntangler
+}
+
+var (
+	serviceInstanceGVK = sc_v1b1.SchemeGroupVersion.WithKind("ServiceInstance")
+)
+
+func WireUp(stateResource *orch_v1.StateResource, context *wiringplugin.WiringContext) (*wiringplugin.WiringResult, bool, error) {
+	err := validateRequest(stateResource, context)
+	if err != nil {
+		return nil, false, err
+	}
+	kubeComputeDependency, err := context.TheOnlyDependency()
+	if err != nil {
+		return nil, false, err
+	}
+	setOfScalingShape, found, err := knownshapes.FindSetOfPScalingShape(kubeComputeDependency.Contract.Shapes)
+	if err != nil {
+		return nil, false, err
+	}
+	if !found {
+		return nil, false, errors.Errorf("failed to find shape %q in contract of %q", knownshapes.SetOfScalingShape, kubeComputeDependency.Name)
+	}
+
+	var wiredResources []smith_v1.Resource
+
+	deploymentResourceName := setOfScalingShape.Data.DeploymentResourceName
+	CPUServiceInstance, err := constructServiceInstance(stateResource, context, deploymentResourceName, CPU)
+	if err != nil {
+		return nil, false, err
+	}
+	wiredResources = append(wiredResources, CPUServiceInstance)
+
+	memoryServiceInstance, err := constructServiceInstance(stateResource, context, deploymentResourceName, Memory)
+	if err != nil {
+		return nil, false, err
+	}
+	wiredResources = append(wiredResources, memoryServiceInstance)
+
+	result := &wiringplugin.WiringResult{
+		Resources: wiredResources,
+	}
+
+	return result, false, nil
+}
+
+func constructServiceInstance(resource *orch_v1.StateResource, context *wiringplugin.WiringContext, deploymentResourceName smith_v1.ResourceName, alarmType AlarmType) (smith_v1.Resource, error) {
+	instanceID, err := svccatentangler.InstanceID(resource.Spec)
+	if err != nil {
+		return smith_v1.Resource{}, err
+	}
+	threshold := AlarmThresholds{
+		Critical: 90,
+		Warning:  80,
+	}
+	query := generateQuerySpec(context, alarmType, threshold)
+	alarmsAtt := createFinalAlarmSpec(resource, &context.StateContext, query)
+	if err != nil {
+		return smith_v1.Resource{}, err
+	}
+	resourceName, err := oap.ResourceName(resource.Spec)
+	if err != nil {
+		return smith_v1.Resource{}, err
+	}
+	if resourceName == "" {
+		resourceName = string(resource.Name)
+	}
+
+	resourceName = resourceName + "--" + string(alarmType)
+	serviceInstanceSpec := ServiceInstanceSpec{
+		ServiceName: context.StateContext.ServiceName,
+		Attributes:  alarmsAtt,
+		Environment: context.StateContext.Location.EnvType,
+		Region:      context.StateContext.Location.Region,
+	}
+
+	serviceInstanceSpecBytes, err := json.Marshal(&serviceInstanceSpec)
+	if err != nil {
+		return smith_v1.Resource{}, err
+	}
+	smithResourceName := voyager.ResourceName(resourceName)
+	alarmInstanceResource := smith_v1.Resource{
+		Name: wiringutil.ServiceInstanceResourceName(smithResourceName),
```

Comment:
This resource name should probably not be constructed from `oap.ResourceName()` because it could potentially clash. I think leaving resource.Name as the prefix is fine

Reply:
Also given none of our other wiring functions produce multiple service instances maybe you can create another function to handle this specific case in wiringutil? `ServiceInstanceWithPostfixResourceName` maybe?